scopesim/effects/data_container.py

```diff
@@ -60,33 +60,56 @@ class DataContainer:
 
     """
 
+    meta = None
+
     def __init__(self, filename=None, table=None, array_dict=None, cmds=None,
                  **kwargs):
 
         self.cmds = cmds
+        # Setting a default for cmds cannot be done here, because from_currsys
+        # checks whether cmds is None. TODO: make this possible.
+        # if self.cmds is None:
+        #     from scopesim import UserCommands
+        #     self.cmds = UserCommands()
+
         if filename is None and "file_name" in kwargs:
             warn("The 'file_name' kwarg is deprecated and will raise an error "
                  "in the future, please use 'filename' instead!",
                  DeprecationWarning, stacklevel=2)
             filename = kwargs["file_name"]
 
+        # A !-string filename is immediately resolved, but the file is not yet
+        # searched for. This makes sense, as the filename should end up in the
+        # FITS headers, and should therefor be independent of any particular
+        # system.
+        #
+        # It would perhaps be even better to not resolve a !-string filename
+        # here, but that currently does not make sense because the file is
+        # directly read in. That is, the file would not be read again if
+        # someone changes the value that the !-string points to. So changing
+        # the value a !-string points to would lead to an inconsistent state.
+        # Immediately resolving the !-string prevents such an inconsistency.
         if isinstance(filename, str) and filename.startswith("!"):
             filename = utils.from_currsys(filename, self.cmds)
 
-        filename = utils.find_file(filename)
-        self.meta = {
+        # A derived clas might have set .meta before calling super().__init__()
+        if self.meta is None:
+            self.meta = {}
+        self.meta.update({
             "filename": filename,
             "description": "",
             "history": [],
             "name": "<empty>",
-        }
+        })
         self.meta.update(kwargs)
 
         self.headers = []
         self.table = None
         self._file = None
 
-        if filename is not None:
+        # Need to check whether the file exists before trying to load it.
+        filename_full = utils.find_file(self.meta["filename"])
+        if filename_full is not None:
```

Comment:
Do we want to throw an Error/Warning/logging when it doesn't exist?

Reply:
`find_file()` logs that warning and raises an Exception if necessary, both optionally. Perhaps finding files needs some refactoring at some point.